easyfsl/methods/relation_networks.py

```diff
@@ -0,0 +1,177 @@
+        maps (in our case the feature map of a query and the feature map of a class prototype).
+        In order to make the network robust to any change in the dimensions of the input images,
+        we made some changes to the architecture defined in the original implementation (typically
+        the use of adaptive pooling).
+        Args:
+            inner_relation_module_channels: number of hidden channels between the linear layers of
+                the relaiton module
+
+        Returns:
+            the constructed relation module
+        """
+        return nn.Sequential(
+            nn.Sequential(
+                nn.Conv2d(
+                    self.feature_dimension * 2,
+                    self.feature_dimension,
+                    kernel_size=3,
+                    padding=1,
+                ),
+                nn.BatchNorm2d(self.feature_dimension, momentum=1, affine=True),
+                nn.ReLU(),
+                nn.AdaptiveMaxPool2d((5, 5)),
+            ),
+            nn.Sequential(
+                nn.Conv2d(
+                    self.feature_dimension,
+                    self.feature_dimension,
+                    kernel_size=3,
+                    padding=0,
+                ),
+                nn.BatchNorm2d(self.feature_dimension, momentum=1, affine=True),
+                nn.ReLU(),
+                nn.AdaptiveMaxPool2d((1, 1)),
+            ),
+            nn.Flatten(),
+            nn.Linear(self.feature_dimension, inner_relation_module_channels),
+            nn.ReLU(),
+            nn.Linear(inner_relation_module_channels, 1),
+            nn.Sigmoid(),
+        )
+
+    def process_support_set(
+        self,
+        support_images: torch.Tensor,
+        support_labels: torch.Tensor,
+    ):
+        """
+        Overwrites process_support_set of AbstractMetaLearner.
+        Extract feature maps from the support set and store class prototypes.
+        """
+
+        support_features = self.backbone(support_images)
+        self.prototypes = compute_prototypes(support_features, support_labels)
+
+    def forward(self, query_images: torch.Tensor) -> torch.Tensor:
+        """
+        Overwrites method forward in AbstractMetaLearner.
+        Predict the label of a query image by concatenating its feature map with each class
+        prototype and feeding the result into a relation module, i.e. a CNN that outputs a relation
+        score. Finally, the classification vector of the query is its relation score to each class
+        prototype.
+        """
+        query_features = self.backbone(query_images)
+
+        # For each pair (query, prototype), we compute the concatenation of their feature maps
+        # Given that query_features is of shape (n_queries, n_channels, width, height), the
+        # constructed tensor is of shape (n_queries * n_prototypes, 2 * n_channels, width, height)
+        query_prototype_feature_pairs = torch.cat(
```

Comment:
Je ne suis pas expert pytorch mais c'est un peu vénère ce redimensionnement. Il y en a pas un plus concis?

Reply:
2 * n_channels since prototype and queries are concatenated